backend_plugin/src/main/evaluators/AddImportEvaluator.java

```diff
@@ -10,102 +10,100 @@
-
-		// If the "t" is the end of an import statement, trigger the feature evaluation
-		return this.checkForTypedImportStatement(docEvent);
-	    }
+	private boolean unresolvedVariablesExist;
+	private boolean lineHadImportStatementAlready;
+
+	/**
+	 * Construct an AddImportEvaluator
+	 */
+	public AddImportEvaluator(IDocument document) {
+		this.featureID = "addImportStatementsSuggestion";
+		this.lineHadImportStatementAlready = false;
+		this.unresolvedVariablesExist = false;
+		this.document = document;
 	}
 
-	// The user did not manually type an import statement
-	return false;
-    }
-
-    /**
-     * Checks the previous 6 characters before the offset of the given DocumentEvent
-     * @param docEvent the DocumentEvent representing the last text insertion
-     * @return boolean true if the previous 6 characters in the document are
-     * 					'i'-'m'-'p'-'o'-'r'-'t'; false otherwise
-     */
-    private boolean checkForTypedImportStatement(DocumentEvent docEvent) {
-	try {
-	    // Check if the six characters before the most recent document change spell "import"
-	    int startOffset = docEvent.getOffset() - 6;
-	    String testing = document.get(startOffset, 6);
-	    if (testing.equals("import")) {
-		// Verify that "import" is at the start of a line
-		int line = document.getLineOfOffset(startOffset);
-		int lineOffset = document.getLineOffset(line);
-		int lineLength = document.getLineLength(line);
-		String lineText = document.get(lineOffset, lineLength).trim();
-		return lineText.startsWith("import");
-	    }
-	} catch (BadLocationException e) {
-	    // The calculated startOffset is negative. This only happens when
-	    // the current offset after the "t " is smaller than 6, meaning
-	    // there is no way the user could have typed "import", so return false
+	/**
+	 * Checks whether a document change event is occurring on a line that did
+	 * not have an import statement previously
+	 * @param event
+	 * @return false
+	 */
+	public boolean evaluateDocumentBeforeChange(DocumentEvent event) {
+		try {
+
+			// If the line before the change event was an import statement,
+			// update the boolean flag to reflect that
+			int line = document.getLineOfOffset(event.getOffset());
+			this.lineHadImportStatementAlready = lineIsAnImportStatement(line);
+		} catch (BadLocationException e) {
+		}
+		return false;
+	}
+
+	/**
+	 * Checks whether a document change event is occurring on a line that did
+	 * not have an import statement previously
+	 * @param event
+	 * @return false
+	 */
+	public boolean evaluateDocumentChanges(DocumentEvent event) {
+		try {
+
+			// If the line before the change was not an import statement, but the
+			// line after the change is, and there are unresolved variables, return true
+			int line = document.getLineOfOffset(event.getOffset());
+			return this.unresolvedVariablesExist && lineIsAnImportStatement(line) &&
+					!this.lineHadImportStatementAlready;
```

Comment:
Small optimization, but move lineIsAnImportStatement(line) to be checked last. That way if this.lineHadImportStatementAlready fails the check, then we skip over doing the more costly method call.

Reply:
Good call, I've updated the PR with the change